File: src/clawsql/api/endpoints/instances.py
"""
Instance management API endpoints.
"""

import asyncio
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status

from ..schemas.requests import (
    InstanceCreateRequest,
    InstanceDiscoverRequest,
    MaintenanceRequest,
    InstanceFilterParams,
    PaginationParams,
)
from ..schemas.responses import (
    DiscoveryResponse,
    InstanceHealthResponse,
    InstanceListResponse,
    InstanceMetricsResponse,
    InstanceResponse,
    SuccessResponse,
)
from ...core.discovery.models import InstanceRole, InstanceState, MySQLInstance
from ...core.discovery.scanner import InstanceRegistry, NetworkScanner

router = APIRouter()

# Global instance registry (would be dependency-injected in production)
_instance_registry: Optional[InstanceRegistry] = None


def get_instance_registry() -> InstanceRegistry:
    """Get or create instance registry."""
    global _instance_registry
    if _instance_registry is None:
        _instance_registry = InstanceRegistry()
    return _instance_registry


@router.get(
    "/",
    response_model=InstanceListResponse,
    summary="List all MySQL instances",
    description="List all discovered and registered MySQL instances with optional filtering.",
)
async def list_instances(
    cluster_id: Optional[str] = Query(None, description="Filter by cluster"),
    state: Optional[str] = Query(None, description="Filter by state"),
    role: Optional[str] = Query(None, description="Filter by role"),
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> InstanceListResponse:
    """
    List all MySQL instances.

    - **cluster_id**: Filter by cluster
    - **state**: Filter by state (online, offline, etc.)
    - **role**: Filter by role (primary, replica)
    """
    instances = registry.get_all()

    # Apply filters
    if cluster_id:
        instances = [i for i in instances if i.cluster_id == cluster_id]
    if state:
        instances = [i for i in instances if i.state.value == state.lower()]
    if role:
        instances = [i for i in instances if i.role.value == role.lower()]

    # Paginate
    total = len(instances)
    start = (page - 1) * page_size
    end = start + page_size
    page_instances = instances[start:end]

    return InstanceListResponse(
        items=[_instance_to_response(i) for i in page_instances],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get(
    "/{instance_id}",
    response_model=InstanceResponse,
    summary="Get instance details",
)
async def get_instance(
    instance_id: str,
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> InstanceResponse:
    """Get details of a specific MySQL instance."""
    instance = registry.get(instance_id)
    if not instance:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Instance not found: {instance_id}",
        )

    return _instance_to_response(instance)


@router.post(
    "/",
    response_model=InstanceResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Register a new instance",
)
async def register_instance(
    request: InstanceCreateRequest,
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> InstanceResponse:
    """
    Manually register a MySQL instance.

    The instance will be validated and added to the registry.
    """
    # Check if already registered
    instance_id = f"{request.host}:{request.port}"
    existing = registry.get(instance_id)
    if existing:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Instance already registered: {instance_id}",
        )

    # Create instance
    instance = MySQLInstance(
        host=request.host,
        port=request.port,
        cluster_id=request.cluster_id,
        labels=request.labels or {},
        state=InstanceState.OFFLINE,
        role=InstanceRole.UNKNOWN,
        last_seen=datetime.utcnow(),
    )

    registry.register(instance)

    return _instance_to_response(instance)


@router.delete(
    "/{instance_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Deregister an instance",
)
async def deregister_instance(
    instance_id: str,
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> None:
    """Remove an instance from management."""
    success = registry.unregister(instance_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Instance not found: {instance_id}",
        )


@router.post(
    "/discover",
    response_model=DiscoveryResponse,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Discover instances in network",
)
async def discover_instances(
    request: InstanceDiscoverRequest,
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> DiscoveryResponse:
    """
    Trigger instance discovery in specified network segments.

    Returns immediately with task ID, results are available via polling.
    """
    import uuid

    task_id = str(uuid.uuid4())

    # Create scanner
    scanner = NetworkScanner(
        network_segments=request.network_segments,
        mysql_port_range=request.port_range or (3306, 3306),
    )

    # Run discovery in background (simplified for demo)
    instances = await scanner.discover_instances()

    # Register discovered instances
    for instance in instances:
        registry.register(instance)

    return DiscoveryResponse(
        task_id=task_id,
        status="completed",
        network_segments=request.network_segments,
        instances_found=len(instances),
        instances=[_instance_to_response(i) for i in instances],
        completed_at=datetime.utcnow(),
    )


@router.post(
    "/{instance_id}/maintenance",
    response_model=SuccessResponse,
    summary="Put instance in maintenance mode",
)
async def set_maintenance(
    instance_id: str,
    request: MaintenanceRequest,
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> SuccessResponse:
    """Put instance in maintenance mode."""
    instance = registry.get(instance_id)
    if not instance:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Instance not found: {instance_id}",
        )

    instance.state = InstanceState.MAINTENANCE
    instance.extra["maintenance_reason"] = request.reason
    instance.extra["maintenance_duration"] = request.duration_minutes

    return SuccessResponse(
        message=f"Instance {instance_id} set to maintenance mode"
    )


@router.delete(
    "/{instance_id}/maintenance",
    response_model=SuccessResponse,
    summary="Remove instance from maintenance",
)
async def remove_maintenance(
    instance_id: str,
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> SuccessResponse:
    """Remove instance from maintenance mode."""
    instance = registry.get(instance_id)
    if not instance:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Instance not found: {instance_id}",
        )

    if instance.state != InstanceState.MAINTENANCE:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Instance is not in maintenance mode: {instance_id}",
        )

    instance.state = InstanceState.ONLINE
    instance.extra.pop("maintenance_reason", None)
    instance.extra.pop("maintenance_duration", None)

    return SuccessResponse(
        message=f"Instance {instance_id} removed from maintenance mode"
    )


@router.get(
    "/{instance_id}/metrics",
    response_model=InstanceMetricsResponse,
    summary="Get instance metrics",
)
async def get_instance_metrics(
    instance_id: str,
    hours: int = Query(1, ge=1, le=24, description="Hours of history"),
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> InstanceMetricsResponse:
    """Get metrics for a specific instance."""
    instance = registry.get(instance_id)
    if not instance:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Instance not found: {instance_id}",
        )

    # Return mock metrics (would come from MetricsCollector in production)
    return InstanceMetricsResponse(
        instance_id=instance_id,
        timestamp=datetime.utcnow(),
        replication_lag_seconds=instance.replication_lag,
        replication_io_running=True,
        replication_sql_running=True,
        connections_current=50,
        connections_max=1000,
        queries_per_second=150.5,
        innodb_buffer_pool_hit_rate=98.5,
        uptime_seconds=864000,
    )


@router.get(
    "/{instance_id}/health",
    response_model=InstanceHealthResponse,
    summary="Get instance health",
)
async def get_instance_health(
    instance_id: str,
    registry: InstanceRegistry = Depends(get_instance_registry),
) -> InstanceHealthResponse:
    """Get current health status of an instance."""
    instance = registry.get(instance_id)
    if not instance:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Instance not found: {instance_id}",
        )

    # Return mock health (would come from HealthChecker in production)
    checks = [
        {
            "check_name": "connectivity",
            "status": "healthy",
            "value": 1.0,
            "message": "Instance is reachable",
        },
        {
            "check_name": "replication_lag",
            "status": "healthy" if not instance.is_replica else "healthy",
            "value": instance.replication_lag or 0.0,
            "message": "Replication lag within threshold",
        },
    ]

    return InstanceHealthResponse(
        instance_id=instance_id,
        status="healthy" if instance.is_online else "unhealthy",
        checks=checks,
    )


def _instance_to_response(instance: MySQLInstance) -> InstanceResponse:
    """Convert MySQLInstance to response schema."""
    return InstanceResponse(
        instance_id=instance.instance_id,
        host=instance.host,
        port=instance.port,
        server_id=instance.server_id,
        role=instance.role.value,
        state=instance.state.value,
        version=instance.version,
        cluster_id=instance.cluster_id,
        replication_lag=instance.replication_lag,
        labels=instance.labels,
        last_seen=instance.last_seen,
        created_at=instance.last_seen,  # Simplified
    )

File: src/api/routes/instances.ts
/**
 * ClawSQL - Instances API Routes
 *
 * Instance management with persistence in shared metadata database.
 * - Topology data (role, state, replication) comes from Orchestrator
 * - User metadata (labels, extra) stored in instance_metadata table
 */

import { FastifyPluginAsync } from 'fastify';
import { v4 as uuidv4 } from 'uuid';
import {
  InstanceRole,
  InstanceState,
  createMySQLInstance,
  createInstanceId,
  MySQLInstance,
} from '../../types/index.js';
import { NotFoundError, AlreadyExistsError } from '../../utils/exceptions.js';
import { getMetricsCollector } from '../../core/monitoring/collector.js';
import { getDatabase } from '../../utils/database.js';
import { getOrchestratorClient } from '../../core/discovery/topology.js';

interface InstanceMetadataRow {
  instance_id: string;
  labels: string;
  extra: string;
  created_at: string;
  updated_at: string;
}

/**
 * Get instance metadata from database
 */
async function getInstanceMetadata(instanceId: string): Promise<InstanceMetadataRow | undefined> {
  const db = getDatabase();
  return db.get<InstanceMetadataRow>(
    'SELECT * FROM instance_metadata WHERE instance_id = ?',
    [instanceId]
  );
}

/**
 * Save instance metadata to database
 */
async function saveInstanceMetadata(
  instanceId: string,
  labels: Record<string, string>,
  extra: Record<string, unknown>
): Promise<void> {
  const db = getDatabase();
  await db.execute(
    `INSERT INTO instance_metadata (instance_id, labels, extra)
     VALUES (?, ?, ?)
     ON DUPLICATE KEY UPDATE labels = VALUES(labels), extra = VALUES(extra)`,
    [instanceId, JSON.stringify(labels), JSON.stringify(extra)]
  );
}

/**
 * Delete instance metadata from database
 */
async function deleteInstanceMetadata(instanceId: string): Promise<void> {
  const db = getDatabase();
  await db.execute('DELETE FROM instance_metadata WHERE instance_id = ?', [instanceId]);
}

/**
 * List all instance metadata from database
 */
async function listInstanceMetadata(): Promise<InstanceMetadataRow[]> {
  const db = getDatabase();
  return db.query<InstanceMetadataRow>('SELECT * FROM instance_metadata');
}

/**
 * Parse JSON field from MySQL (can be object or string)
 */
function parseJsonField(value: unknown): Record<string, unknown> {
  if (!value) return {};
  if (typeof value === 'object') return value as Record<string, unknown>;
  if (typeof value === 'string') {
    try {
      return JSON.parse(value);
    } catch {
      return {};
    }
  }
  return {};
}

/**
 * Parse labels field (string values only)
 */
function parseLabels(value: unknown): Record<string, string> {
  const parsed = parseJsonField(value);
  const result: Record<string, string> = {};
  for (const [key, val] of Object.entries(parsed)) {
    if (typeof val === 'string') {
      result[key] = val;
    }
  }
  return result;
}

/**
 * Convert to API response
 */
function toResponse(
  metadata: InstanceMetadataRow,
  orchestratorData?: MySQLInstance | null
) {
  const labels = parseLabels(metadata.labels);
  const extra = parseJsonField(metadata.extra);

  // Parse instance_id to get host and port
  const [host, portStr] = metadata.instance_id.split(':');
  const port = parseInt(portStr, 10) || 3306;

  // Convert InstanceRole enum to string
  const role = orchestratorData?.role?.toLowerCase() || 'unknown';
  const state = orchestratorData?.state?.toLowerCase() || 'offline';

  return {
    instance_id: metadata.instance_id,
    host,
    port,
    server_id: orchestratorData?.serverId ?? null,
    role,
    state,
    version: orchestratorData?.version ?? null,
    cluster_id: orchestratorData?.clusterId ?? null,
    replication_lag: orchestratorData?.replicationLag ?? null,
    labels,
    extra,
    last_seen: orchestratorData?.lastSeen?.toISOString?.() ?? metadata.updated_at,
    created_at: metadata.created_at,
  };
}

const instancesRoutes: FastifyPluginAsync = async (fastify) => {
  // List instances
  fastify.get<{
    Querystring: {
      cluster_id?: string;
      state?: string;
      role?: string;
      page?: number;
      page_size?: number;
    };
  }>('/', async (request) => {
    const metadataList = await listInstanceMetadata();
    const orchestrator = getOrchestratorClient();

    // Get topology from Orchestrator for enrichment
    let instances = await Promise.all(
      metadataList.map(async (m) => {
        const [host, portStr] = m.instance_id.split(':');
        const port = parseInt(portStr, 10) || 3306;
        try {
          const topo = await orchestrator.getInstance(host, port);
          return toResponse(m, topo);
        } catch {
          return toResponse(m);
        }
      })
    );

    // Apply filters
    if (request.query.cluster_id) {
      instances = instances.filter(i => i.cluster_id === request.query.cluster_id);
    }
    if (request.query.state) {
      instances = instances.filter(i => i.state === request.query.state);
    }
    if (request.query.role) {
      instances = instances.filter(i => i.role === request.query.role);
    }

    // Paginate
    const page = request.query.page || 1;
    const pageSize = request.query.page_size || 20;
    const total = instances.length;
    const start = (page - 1) * pageSize;
    const end = start + pageSize;

    return {
      items: instances.slice(start, end),
      total,
      page,
      page_size: pageSize,
    };
  });

  // Get instance
  fastify.get<{
    Params: { instance_id: string };
  }>('/:instance_id', async (request) => {
    const metadata = await getInstanceMetadata(request.params.instance_id);
    if (!metadata) {
      throw new NotFoundError('Instance', request.params.instance_id);
    }

    const [host, portStr] = request.params.instance_id.split(':');
    const port = parseInt(portStr, 10) || 3306;
    const orchestrator = getOrchestratorClient();

    try {
      const topo = await orchestrator.getInstance(host, port);
      return toResponse(metadata, topo);
    } catch {
      return toResponse(metadata);
    }
  });

  // Register instance
  fastify.post<{
    Body: {
      host: string;
      port?: number;
      cluster_id?: string;
      labels?: Record<string, string>;
    };
  }>('/', async (request, reply) => {
    const port = request.body.port || 3306;
    const instanceId = createInstanceId(request.body.host, port);

    const existing = await getInstanceMetadata(instanceId);
    if (existing) {
      throw new AlreadyExistsError('Instance', instanceId);
    }

    await saveInstanceMetadata(instanceId, request.body.labels || {}, {
      registeredCluster: request.body.cluster_id,
    });

    // Also discover in Orchestrator
    const orchestrator = getOrchestratorClient();
    try {
      await orchestrator.discoverInstance(request.body.host, port);
    } catch (error) {
      fastify.log.warn({ error, instanceId }, 'Failed to discover instance in Orchestrator');
    }

    const metadata = await getInstanceMetadata(instanceId);
    reply.code(201);
    return toResponse(metadata!);
  });

  // Deregister instance
  fastify.delete<{
    Params: { instance_id: string };
  }>('/:instance_id', async (request, reply) => {
    const existing = await getInstanceMetadata(request.params.instance_id);
    if (!existing) {
      throw new NotFoundError('Instance', request.params.instance_id);
    }

    await deleteInstanceMetadata(request.params.instance_id);
    reply.code(204);
  });

  // Discover instances
  fastify.post<{
    Body: {
      network_segments: string[];
      port_range?: [number, number];
    };
  }>('/discover', async (request, reply) => {
    const taskId = uuidv4();

    // Simplified discovery - in production this would scan networks
    const discovered: { instance_id: string; host: string; port: number }[] = [];

    for (const segment of request.body.network_segments) {
      fastify.log.info({ segment }, 'Scanning network segment');
    }

    // Register discovered instances
    for (const instance of discovered) {
      const instanceId = createInstanceId(instance.host, instance.port);
      const existing = await getInstanceMetadata(instanceId);
      if (!existing) {
        await saveInstanceMetadata(instanceId, {}, { discovered: true });
      }
    }

    reply.code(202);
    return {
      task_id: taskId,
      status: 'completed',
      network_segments: request.body.network_segments,
      instances_found: discovered.length,
      instances: discovered,
      completed_at: new Date().toISOString(),
    };
  });

  // Set maintenance mode
  fastify.post<{
    Params: { instance_id: string };
    Body: { reason: string; duration_minutes?: number };
  }>('/:instance_id/maintenance', async (request) => {
    const metadata = await getInstanceMetadata(request.params.instance_id);
    if (!metadata) {
      throw new NotFoundError('Instance', request.params.instance_id);
    }

    const extra = parseJsonField(metadata.extra);
    extra.maintenance_reason = request.body.reason;
    extra.maintenance_duration = request.body.duration_minutes || 60;
    extra.maintenance_mode = true;

    const labels = parseLabels(metadata.labels);
    await saveInstanceMetadata(request.params.instance_id, labels, extra);

    return { message: `Instance ${request.params.instance_id} set to maintenance mode` };
  });

  // Remove maintenance mode
  fastify.delete<{
    Params: { instance_id: string };
  }>('/:instance_id/maintenance', async (request, reply) => {
    const metadata = await getInstanceMetadata(request.params.instance_id);
    if (!metadata) {
      throw new NotFoundError('Instance', request.params.instance_id);
    }

    const extra = parseJsonField(metadata.extra);
    if (!extra.maintenance_mode) {
      return reply.code(400).send({
        error: 'INVALID_STATE',
        message: `Instance is not in maintenance mode: ${request.params.instance_id}`,
      });
    }

    delete extra.maintenance_reason;
    delete extra.maintenance_duration;
    delete extra.maintenance_mode;

    const labels = parseLabels(metadata.labels);
    await saveInstanceMetadata(request.params.instance_id, labels, extra);

    return { message: `Instance ${request.params.instance_id} removed from maintenance mode` };
  });

  // Get instance metrics
  fastify.get<{
    Params: { instance_id: string };
  }>('/:instance_id/metrics', async (request) => {
    const metadata = await getInstanceMetadata(request.params.instance_id);
    if (!metadata) {
      throw new NotFoundError('Instance', request.params.instance_id);
    }

    const [host, portStr] = request.params.instance_id.split(':');
    const port = parseInt(portStr, 10) || 3306;

    const instance = createMySQLInstance(host, port, {
      state: InstanceState.ONLINE,
      role: InstanceRole.UNKNOWN,
    });

    const collector = getMetricsCollector();
    const metrics = await collector.collectMetrics(instance);

    return {
      ...metrics,
      timestamp: metrics.timestamp.toISOString(),
    };
  });

  // Get instance health
  fastify.get<{
    Params: { instance_id: string };
  }>('/:instance_id/health', async (request) => {
    const metadata = await getInstanceMetadata(request.params.instance_id);
    if (!metadata) {
      throw new NotFoundError('Instance', request.params.instance_id);
    }

    const [host, portStr] = request.params.instance_id.split(':');
    const port = parseInt(portStr, 10) || 3306;
    const orchestrator = getOrchestratorClient();

    let isOnline = false;
    let replicationLag: number | undefined;

    try {
      const topo = await orchestrator.getInstance(host, port);
      isOnline = true;
      replicationLag = (topo as unknown as { replication_lag_seconds?: number }).replication_lag_seconds;
    } catch {
      isOnline = false;
    }

    const extra = parseJsonField(metadata.extra);
    const inMaintenance = extra.maintenance_mode === true;

    const checks = [
      {
        check_name: 'connectivity',
        status: inMaintenance ? 'maintenance' : (isOnline ? 'healthy' : 'unhealthy'),
        value: isOnline ? 1 : 0,
        message: inMaintenance ? 'Instance is in maintenance mode' : (isOnline ? 'Instance is reachable' : 'Instance is not reachable'),
      },
      {
        check_name: 'replication_lag',
        status: replicationLag !== undefined && replicationLag < 10 ? 'healthy' : 'unhealthy',
        value: replicationLag ?? 0,
        message: replicationLag !== undefined
          ? `Replication lag: ${replicationLag}s`
          : 'Replication not configured',
      },
    ];

    return {
      instance_id: request.params.instance_id,
      status: inMaintenance ? 'maintenance' : (isOnline ? 'healthy' : 'unhealthy'),
      checks,
    };
  });
};

export default instancesRoutes;